File: java/rooms/RoomBuilderImpl.java
package rooms;

import chest.Chest;
import chest.ChestFactory;
import chest.ChestType;
import monsters.CreatureFactory;
import monsters.CreatureType;
import monsters.options.Creature;
import rooms.options.GameMasterable;
import rooms.options.RoomType;
import rooms.options.RoomWithChest;
import rooms.options.RoomWithCreature;

public class RoomBuilderImpl implements RoomBuilder {
    private final RoomFactory roomFactory;
    private final CreatureFactory creatureFactory;
    private final ChestFactory chestFactory;

    private GameMasterable room;

    public RoomBuilderImpl(RoomFactory roomFactory, CreatureFactory creatureFactory, ChestFactory chestFactory) {
        this.roomFactory = roomFactory;
        this.creatureFactory = creatureFactory;
        this.chestFactory = chestFactory;
    }
    public void createEmptyRoom(RoomType type) {
        room = roomFactory.createEmptyRoom(type);
    }
    @Override
    public void addCreature(CreatureType type) {
        if(room == null) return;
        Creature creature = creatureFactory.createCreatureFromType(type);
        room = new RoomWithCreature(room, creature);
    }

    @Override
    public void addChest(ChestType type) {
        if(room == null) return;
        Chest chest = chestFactory.createChestByType(type);
        room = new RoomWithChest(room, chest);
    }

    @Override
    public void reset() {
        room = null;
    }

    @Override
    public GameMasterable build() {
        return room;
    }
}

File: php/rooms/RoomBuilderImpl.php
<?php

namespace rooms;

use chest\Chest;
use chest\ChestFactory;
use chest\ChestType;
use monsters\CreatureFactory;
use monsters\CreatureType;
use monsters\options\Creature;
use rooms\options\GameMasterable;
use rooms\options\RoomType;
use rooms\options\RoomWithChest;
use rooms\options\RoomWithCreature;

class RoomBuilderImpl implements RoomBuilder {
    private $roomFactory;
    private $creatureFactory;
    private $chestFactory;

    private $room;

    public function __construct($roomFactory, $creatureFactory, $chestFactory) {
        $this->roomFactory = $roomFactory;
        $this->creatureFactory = $creatureFactory;
        $this->chestFactory = $chestFactory;
    }

    public function createEmptyRoom($type) {
        $this->room = $this->roomFactory->createEmptyRoom($type);
    }

    public function addCreature($type) {
        if ($this->room === null) return;
        $creature = $this->creatureFactory->createCreatureFromType($type);
        $this->room = new RoomWithCreature($this->room, $creature);
    }

    public function addChest($type) {
        if ($this->room === null) return;
        $chest = $this->chestFactory->createChestByType($type);
        $this->room = new RoomWithChest($this->room, $chest);
    }

    public function reset() {
        $this->room = null;
    }

    public function build() {
        return $this->room;
    }
}
?>